api/BUILD

```diff
@@ -7,12 +9,39 @@ api_proto_library(
     srcs = ["address.proto"],
 )
 
+go_proto_library(
+    name = "address_go_proto",
+    importpath = "github.com/envoyproxy/data-plane-api/api/address",
+    proto = ":address",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@com_github_gogo_protobuf//:gogo_proto_go",
+        "@com_github_golang_protobuf//ptypes/wrappers:go_default_library",
+        "@com_lyft_protoc_gen_validate//validate:go_default_library",
+    ],
+)
+
 api_proto_library(
     name = "base",
     srcs = ["base.proto"],
     deps = [":address"],
 )
 
+go_proto_library(
```

Comment:
Could we have the `api_proto_library` macro also define the `go_proto_library`? Seems most of the complexity there would be around `deps`, but I think there is a fairly mechanical transform from the existing `api_proto_library` `deps` to the `go_proto_library` ones.

Reply:
A consequence of this is that all people using api_proto_library will depend on Go rules whether they need them or not.

Given the immaturity of Go bazel rules are we sure we want this coupling? This will force people developing updates to the protos who are only familiar with C++ to need to start grappling with Go bazel rules which is pretty brittle and complicated right now.

Note that I have not created Go gRPC rules for metrics_service.proto because I can't figure out how to depend on the Go version of @promotheus_metrics_model//:client_model. And until I figure out how to do that, I can't actually make api_proto_library also define the go_proto_library.

Someone else is having a closely related problem:
https://github.com/envoyproxy/data-plane-api/issues/231

kyessenov's solution to use an import map doesn't work if you want to use Bazel Go rules since Bazel Go rules don't allow you to specify an import map yet.

If you are good at protoc here are some system output that may be helpful

* protoc subcommand
cd /private/var/tmp/_bazel_misterwilliam/e587f3ce7c8cd4619820a54b06276170/execroot/envoy_api && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/darwin_amd64_stripped/compile -go bazel-out/host/bin/external/go_stdlib_darwin_amd64_cgo/bin/go -root_file bazel-out/host/bin/external/go_stdlib_darwin_amd64_cgo/ROOT -goos darwin -goarch amd64 '-cgo=1' -package_list external/go_sdk/packages.txt -src 'bazel-out/darwin-fastbuild/bin/api/metrics_go_grpc~/github.com/envoyproxy/data-plane-api/api/metrics/metrics_service.pb.go' -dep github.com/envoyproxy/data-plane-api/api/base -dep google.golang.org/genproto/googleapis/api/annotations -dep 'github.com/promotheus_metrics_model/io_prometheus_client;.' -dep github.com/golang/protobuf/proto -dep google.golang.org/grpc -dep golang.org/x/net/context -I 'bazel-out/darwin-fastbuild/bin/api/darwin_amd64_stripped/base_go_proto~/' -I 'bazel-out/darwin-fastbuild/bin/external/googleapis/darwin_amd64_stripped/http_api_go_proto~/' -I 'bazel-out/darwin-fastbuild/bin/external/promotheus_metrics_model/darwin_amd64_stripped/client_model_go_proto~/' -I 'bazel-out/darwin-fastbuild/bin/external/com_github_golang_protobuf/proto/darwin_amd64_stripped/go_default_library~/' -I 'bazel-out/darwin-fastbuild/bin/external/org_golang_google_grpc/darwin_amd64_stripped/go_default_library~/' -I 'bazel-out/darwin-fastbuild/bin/external/org_golang_x_net/context/darwin_amd64_stripped/go_default_library~/' -o 'bazel-out/darwin-fastbuild/bin/api/darwin_amd64_stripped/metrics_go_grpc~/github.com/envoyproxy/data-plane-api/api/metrics.a' -trimpath . -I . -- -p github.com/envoyproxy/data-plane-api/api/metrics)
ERROR: /Users/misterwilliam/Workspace/data-plane-api/api/BUILD:242:1: GoCompile api/darwin_amd64_stripped/metrics_go_grpc~/github.com/envoyproxy/data-plane-api/api/metrics.a failed (Exit 1)
2017/12/24 17:56:16 missing strict dependencies:
	bazel-out/darwin-fastbuild/bin/api/metrics_go_grpc~/github.com/envoyproxy/data-plane-api/api/metrics/metrics_service.pb.go: import of ., which is not a direct dependency

* protoc output snippet of metrics_service.pb.go
```
import io_prometheus_client "."
```

* How I tried to import prometheus in repositories.bzl
```
    native.new_http_archive(
        name = "promotheus_metrics_model",
        strip_prefix = "client_model-" + PROMETHEUS_SHA,
        url = "https://github.com/prometheus/client_model/archive/" + PROMETHEUS_SHA + ".tar.gz",
        build_file_content = """
load("@envoy_api//bazel:api_build_system.bzl", "api_proto_library")
load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")

api_proto_library(
    name = "client_model",
    srcs = [
        "metrics.proto",
    ],
)

go_proto_library(
  name = "client_model_go_proto",
  importpath = "github.com/promotheus_metrics_model/io_prometheus_client;.",
  proto = ":client_model",
  visibility = ["//visibility:public"],
  deps = [
  ],
)
        """,
```

Reply:
OK, I think it's fine to decouple as you have, but please add a tracking issues and some TODOs in the `.bzl` or `BUILD` to keep in mind that we want to converge at some point.

Reply:
I'm guessing your `protoc` issue is some kind of proto path one, e.g. `option go_package` stuff, see also https://github.com/golang/protobuf/issues/139.

Reply:
Created issue: https://github.com/envoyproxy/data-plane-api/issues/373

I added the TODO api_build_system.bzl, and assigned it to you. Is that okay?

(My gut feeling is that automatically generating go_proto_library rules from api_proto_library is a bad idea though. It forces everyone to depend on proto libraries in all languages whether people want them or not. Note that internally Google's proto libraries used to be done very similar api_proto_library, and we migrated away from that because of forcing people to depend on proto's for languages they didn't want was considered undesirable.)

Reply:
Ideally we would have all dependencies captured in the `proto_library` target and the language specifies would just be an aspect that runs over this, so we wouldn't need to have duplicate boiler plate stuff here. I'm guessing `go_proto_library` isn't quite smart enough to do this yet, but that would be the ideal long term direction.